File: codes/202603/26/SentimentAnalyzer.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace NewsSentimentAnalyzer
{
    public class SentimentAnalyzer
    {
        private readonly HashSet<string> _positiveWords;
        private readonly HashSet<string> _negativeWords;

        public SentimentAnalyzer()
        {
            // 간단한 긍정 및 부정 단어 사전 (한국어)
            // 실제 프로젝트에서는 훨씬 더 크고 정교한 사전을 사용합니다.
            _positiveWords = new HashSet<string>
            {
                "좋다", "긍정적", "행복", "기쁘다", "성공", "발전", "희망", "탁월", "뛰어나다", "칭찬",
                "환영", "개선", "성장", "강화", "지원", "해결", "기대", "도움", "안정", "회복"
            };

            _negativeWords = new HashSet<string>
            {
                "나쁘다", "부정적", "슬픔", "화나다", "실패", "문제", "우려", "악화", "비판", "불만",
                "논란", "축소", "약화", "반대", "어렵다", "위험", "부담", "침체", "하락", "갈등"
            };
        }

        public string AnalyzeSentiment(string text)
        {
            if (string.IsNullOrWhiteSpace(text))
            {
                return "분석할 텍스트가 없습니다.";
            }

            // 텍스트를 소문자로 변환하고 단어 단위로 분리
            // 간단화를 위해 구두점은 무시하고 공백으로만 분리합니다.
            var words = text.ToLowerInvariant()
                            .Split(new[] { ' ', ',', '.', '!', '?', ';' }, StringSplitOptions.RemoveEmptyEntries);

            int positiveScore = 0;
            int negativeScore = 0;

            foreach (var word in words)
            {
                if (_positiveWords.Contains(word))
                {
                    positiveScore++;
                }
                else if (_negativeWords.Contains(word))
                {
                    negativeScore++;
                }
            }

            if (positiveScore > negativeScore)
            {
                return "긍정적";
            }
            else if (negativeScore > positiveScore)
            {
                return "부정적";
            }
            else
            {
                return "중립적";
            }
        }
    }
}

File: codes/202604/15/com/gemini/sentiment/SentimentAnalyzer.java
// SentimentAnalyzer.java
// 문장의 감성을 분석하는 핵심 로직을 담은 클래스 (산업용 알고리즘 개념 활용)

package com.gemini.sentiment;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class SentimentAnalyzer {

    // 긍정적인 단어 집합 (산업용 라이브러리/데이터베이스에서 가져온다고 가정)
    private static final Set<String> POSITIVE_WORDS = new HashSet<>(Arrays.asList(
            "좋다", "최고", "훌륭하다", "만족", "기쁘다", "행복", "사랑", "성공", "매우 좋다", "아름답다"
    ));

    // 부정적인 단어 집합 (산업용 라이브러리/데이터베이스에서 가져온다고 가정)
    private static final Set<String> NEGATIVE_WORDS = new HashSet<>(Arrays.asList(
            "나쁘다", "최악", "끔찍하다", "불만", "슬프다", "고통", "실패", "문제", "싫다", "역겹다"
    ));

    public SentimentAnalyzer() {
        // 실제 시스템에서는 감성 사전 또는 머신러닝 모델을 로드하는 로직이 들어갈 수 있습니다.
        System.out.println("감성 분석기 초기화 중...");
    }

    /**
     * 주어진 문장의 감성을 분석합니다.
     * 이 메서드는 실제 산업에서 사용되는 복잡한 NLP(자연어 처리) 알고리즘의
     * 간소화된 버전을 나타냅니다. 여기서는 키워드 매칭을 사용하지만,
     * 실제로는 통계적 모델(예: 나이브 베이즈), 머신러닝 모델(예: SVM, 딥러닝),
     * 또는 렉시콘 기반(감성 사전) 접근 방식이 사용됩니다.
     *
     * @param sentence 분석할 문장 객체
     * @return 감성이 설정된 문장 객체
     */
    public Sentence analyze(Sentence sentence) {
        String text = sentence.getText().toLowerCase(); // 소문자로 변환하여 분석 효율성 증대
        int positiveScore = 0;
        int negativeScore = 0;

        // 문장을 단어 단위로 분리 (간단한 공백 기준 분리, 실제로는 형태소 분석기가 필요)
        String[] words = text.split("\s+");

        for (String word : words) {
            // 단어의 불필요한 구두점 제거 (예: 마침표, 쉼표 등)
            word = word.replaceAll("[.,!?]", "");

            if (POSITIVE_WORDS.contains(word)) {
                positiveScore++;
            } else if (NEGATIVE_WORDS.contains(word)) {
                negativeScore++;
            }
        }

        if (positiveScore > negativeScore) {
            sentence.setSentiment(Sentiment.POSITIVE);
        } else if (negativeScore > positiveScore) {
            sentence.setSentiment(Sentiment.NEGATIVE);
        } else {
            sentence.setSentiment(Sentiment.NEUTRAL); // 점수가 같거나 키워드가 없으면 중립
        }
        return sentence;
    }
}
